File: frameworks/persistence.spring/src/main/java/com/example/scaffold/frameworks/persistence/spring/SpringPage.java
package com.example.scaffold.frameworks.persistence.spring;

import com.example.frameworks.domain.core.Page;
import com.example.frameworks.domain.core.Sort;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.function.Function;

public class SpringPage<T> implements Page<T> {

    private final org.springframework.data.domain.Page<T> impl;

    public static <T> SpringPage<T> from(org.springframework.data.domain.Page<T> page) {
        return new SpringPage<>(page);
    }

    protected SpringPage(org.springframework.data.domain.Page<T> impl) {
        this.impl = impl;
    }

    @Override
    public @NotNull List<T> content() {
        return impl.getContent();
    }

    @Override
    public int size() {
        return impl.getSize();
    }

    @Override
    public int number() {
        return impl.getNumber();
    }

    @Override
    public int numberOfElements() {
        return impl.getNumberOfElements();
    }

    @Override
    public int numberOfTotalPages() {
        return impl.getTotalPages();
    }

    @Override
    public long numberOfTotalElements() {
        return impl.getTotalElements();
    }

    @Override
    public boolean first() {
        return impl.isFirst();
    }

    @Override
    public boolean last() {
        return impl.isLast();
    }

    @Override
    public @NotNull Sort sort() {
        return SpringSort.of(impl.getSort());
    }

    @Override
    public @NotNull <U> Page<U> map(Function<? super T, ? extends U> converter) {
        return new SpringPage<>(impl.map(converter));
    }

}

File: frameworks/persistence.spring/src/main/kotlin/com/qomolangma/frameworks/persistence/spring/SpringPage.kt
package com.qomolangma.frameworks.persistence.spring

import com.qomolangma.frameworks.domain.core.Sort
import org.springframework.data.domain.Page
import java.util.function.Function

open class SpringPage<T> protected constructor(private val impl: Page<T>) :
    com.qomolangma.frameworks.domain.core.Page<T> {
    override fun content(): List<T> {
        return impl.content
    }

    override fun size(): Int {
        return impl.size
    }

    override fun number(): Int {
        return impl.number
    }

    override fun numberOfElements(): Int {
        return impl.numberOfElements
    }

    override fun numberOfTotalPages(): Int {
        return impl.totalPages
    }

    override fun numberOfTotalElements(): Long {
        return impl.totalElements
    }

    override fun first(): Boolean {
        return impl.isFirst
    }

    override fun last(): Boolean {
        return impl.isLast
    }

    override fun sort(): Sort {
        return SpringSort.of(impl.sort)
    }

    override fun <U> map(converter: Function<in T, out U>): com.qomolangma.frameworks.domain.core.Page<U> {
        return SpringPage(impl.map(converter))
    }

    companion object {
        @JvmStatic
        fun <T> from(page: Page<T>): SpringPage<T> {
            return SpringPage(page)
        }
    }
}
